nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,630 @@
+
+        // go to a known state
+        radio.disable();
+
+        // clear any event interesting to us
+        radio.radio.events_disabled.reset();
+        radio.radio.events_end.reset();
+
+        radio.radio.mode.write(|w| w.mode().ieee802154_250kbit());
+
+        // NOTE(unsafe) radio is currently disabled
+        unsafe {
+            radio.radio.pcnf0.write(|w| {
+                w.s1incl()
+                    .clear_bit() // S1 not included in RAM
+                    .plen()
+                    ._32bit_zero() // 32-bit zero preamble
+                    .crcinc()
+                    .include() // LENGTH field includes CRC (2 bytes)
+                    .cilen()
+                    .bits(0) // no code indicator
+                    .lflen()
+                    .bits(8) // length = 8 bits (but highest bit is reserved and must be `0`)
+                    .s0len()
+                    .clear_bit() // no S0
+                    .s1len()
+                    .bits(0) // no S1
+            });
+
+            radio.radio.pcnf1.write(|w| {
+                w.maxlen()
+                    .bits(Packet::MAX_LEN + 2 /* CRC */) // payload length
+                    .statlen()
+                    .bits(0) // no static length
+                    .balen()
+                    .bits(0) // no base address
+                    .endian()
+                    .clear_bit() // little endian
+                    .whiteen()
+                    .clear_bit() // no data whitening
+            });
+
+            // CRC configuration required by the IEEE spec: x**16 + x**12 + x**5 + 1
+            radio
+                .radio
+                .crccnf
+                .write(|w| w.len().two().skipaddr().ieee802154());
+            radio.radio.crcpoly.write(|w| w.crcpoly().bits(0x11021));
+            radio.radio.crcinit.write(|w| w.crcinit().bits(0));
+
+            // permanent shortcuts
+            radio.radio.shorts.write(|w| w.txready_start().set_bit());
+        }
+
+        // set default settings
+        radio.set_cca(DEFAULT_CCA);
+        radio.set_channel(DEFAULT_CHANNEL);
+        radio.set_sfd(DEFAULT_SFD);
+        radio.set_txpower(DEFAULT_TXPOWER);
+
+        radio
+    }
+
+    /// Changes the radio channel
+    pub fn set_channel(&mut self, channel: Channel) {
+        self.disable();
+
+        // NOTE(unsafe) radio is currently disabled
+        unsafe {
+            self.radio
+                .frequency
+                .write(|w| w.map().clear_bit().frequency().bits(channel as u8))
+        }
+    }
+
+    /// Changes the Clear Channel Assessment method
+    pub fn set_cca(&mut self, cca: Cca) {
+        self.disable();
+
+        match cca {
+            Cca::CarrierSense => self.radio.ccactrl.write(|w| w.ccamode().carrier_mode()),
+        }
+    }
+
+    /// Changes the Start of Frame Delimiter
+    pub fn set_sfd(&mut self, sfd: u8) {
+        // FIXME don't completely turn off the radio; RXIDLE or TXIDLE are probably OK
```

Comment:
I expect that every state is OK, I think this register is also double buffered, but I'm not sure about it, because we have never modified SFD value.